crates/typed-store/src/metrics.rs

```diff
@@ -1,12 +1,72 @@
 // Copyright (c) 2022, Mysten Labs, Inc.
 // SPDX-License-Identifier: Apache-2.0
-use std::sync::Arc;
-
 use once_cell::sync::OnceCell;
 use prometheus::{
     exponential_buckets, register_histogram_vec_with_registry,
     register_int_counter_vec_with_registry, HistogramVec, IntCounterVec, Registry,
 };
+use std::sync::atomic::{AtomicU64, Ordering};
+use std::sync::Arc;
+use std::time::Duration;
+
+#[derive(Debug, Clone)]
+// A struct for sampling based on number of operations or duration.
+// Sampling happens if the duration expires and after number of operations
+pub struct SamplingInterval {
+    // Sample once every time duration
+    pub once_every_duration: Duration,
+    // Sample once every number of operations
+    pub after_num_ops: u64,
+    // Counter for keeping track of previous sample
+    pub counter: Arc<AtomicU64>,
+}
+
+impl Default for SamplingInterval {
+    fn default() -> Self {
+        SamplingInterval::new(Duration::from_secs(1), 0)
+    }
+}
+
+impl SamplingInterval {
+    pub fn new(once_every_duration: Duration, after_num_ops: u64) -> Self {
+        let counter = if !once_every_duration.is_zero() {
+            // start with counter set to 1 instead of 0 so first check of
+            // sample() does not return true
+            let counter = Arc::new(AtomicU64::new(1));
+            let counter_cloned = counter.clone();
+            tokio::task::spawn(async move {
+                let counter = counter_cloned.clone();
+                loop {
+                    let counter = counter.clone();
+                    if counter.load(Ordering::SeqCst) > after_num_ops {
+                        counter.store(0, Ordering::SeqCst);
+                    }
+                    tokio::time::sleep(once_every_duration).await;
+                }
+            });
+            counter
+        } else {
+            Arc::new(AtomicU64::new(0))
+        };
+        SamplingInterval {
+            once_every_duration,
+            after_num_ops,
+            counter,
+        }
+    }
+    pub fn sample(&self) -> bool {
+        if self.once_every_duration.is_zero() {
+            if self.counter.fetch_add(1, Ordering::Relaxed) == self.after_num_ops {
```

Comment:
You can also do
```
self.counter.fetch_add(1, Ordering::Relaxed)  % self.after_num_ops == 0
```
(and then don't do `self.counter.store(0)`)
This would be more precise and avoid race conditions on the counter

Reply:
hmm, that could technically overflow the counter if we never reset it

Reply:
I think it is not practically likely that we overflow u64 counter.